### PR TITLE
Revised treatment of alternate stacks for signal handing

### PR DESCRIPTION
**Pull request by xavierleroy**
This PR brings alternate signal stack handling in OCaml 5 up to date with what's in 4.14 or was proposed for 4.14:

* First commit makes sure that signal handlers are run on the alternate signal stack.  This is required since fiber stacks are generally too small to run signal handlers on.

* Second commit makes sure the alternate signal stack for domain 0 is turned off and freed when the main OCaml code stops. (Adaptation of #10726)
    
* Second commit also avoids a crash when other C libraries install their own alternate signal stack.  Don't try to free their stack, only the stack block we allocated. (Adaptation of #11496)

* Third commit makes sure that every thread created with `Thread.create` has its own alternate signal stack.  Currently, all threads created by a domain race to use the signal stack of this domain.

* Fourth commit (added later) moves the tweaking of the SIGPROF handler from "every time a domain or thread is created" to "once when the process starts", which is enough.

Cc @kayceesrk @Engil .

### CODE REVIEW

**Comment by xavierleroy** [comment]
Note to reviewers: when the main program is in C and the OCaml code is manually started using `caml_startup`, it's not obvious where to call `caml_terminate_signals`.  In this PR, like in 4.14, it's called when `caml_startup` returns.  However, as pointed out in  #11486, the OCaml code may have installed callbacks that, when run, will need OCaml's special signal handling.  This will have to be discussed separately.  This PR just tries to get OCaml 5 to parity with OCaml 4.14.

**Comment by xavierleroy** [comment]
Something mysterious is happening... 

`testsuite/tests/asmcomp/polling_insertion.ml` fails the assertion at line 253, but only(?) with flambda, and only(?) on x86-64 (for once, PowerPC and S390x are fine).  Yet the assertion is about requesting minor GCs and checking when the requests are honored; nothing to do with alternate signal stacks, at first sight.

Could someone who understands flambda and polling points help me understand what's going on?  @lthls or @sadiqj perhaps?

**Comment by lthls** [comment]
I had a quick look, and it seems that the poll point at the start of the function triggers a GC as expected, but for some reason this GC increments the minor GC stats twice (once during `caml_empty_minor_heaps_once`, and once during `caml_major_collection_slice`; both functions are called in `caml_poll_gc_work`). I don't know yet what are the conditions for this to happen.

File: runtime/startup_byt.c
Comment (gasche) [inline runtime/startup_byt.c:592]: Don't we need to also cleanup signals in the uncaught-exception case above as well?

**Comment by gasche** [comment]
On the code of the PR itself:

- Signal stacks are handled unlike other domain- or thread-local resources, they are initialized in the callback or function running the domain or thread code and cleaned in each exit path of this callback (hopefully).
- For threads and non-0 domains, this is easy to do -- but it stands out as a completely different way to do resource initialization and cleanup.
- For domain 0 it is less simple, the resource is tracked by a new global variable to pass it from its initialization to its cleanup point.

I think the code would be simpler to explain (one less special case) if the signal stack was stored in the usual domain_state and thread-local structures, without a special case with a different cleanup logic. However,
1. This requires a bit more boilerplate to pass around information that we don't use directly during OCaml code execution.
2. Presently no effort was done in the runtime code to cleanup resources from domain 0 on exit, so doing this properly would require a refactoring. (I don't know if we would want to cleanup all resources, or only some of them, possibly only this one.)

This suggests that the present PR is one possible solution to a compromise. With my current understanding it's not my favorite solution (it adds a special case that makes the code harder to explain), but my favorite solution requires more, independent work, so it's sensible to go ahead with this as a first step.

One aspect of this PR that is a bit fragile is the necessity to add new cleanup code on all exit points for the main domain. Other domains, and threads have clean code with a single exit point, but not so for the main domain in the cases of errors. I am not sure that all exit points have been covered.

**Comment by abbysmal** [comment]
I am currently on leave (Tarides is closed until Monday), but I will take a look at this PR when we re-open on Monday, sorry for the delay answering the ping. :)

**Comment by xavierleroy** [comment]
Re: Tarides currently closed, please take your time, I'll be traveling and mostly unavailable next week myself.

Re: storing the signal stack pointer in the domain state, that was my first attempt, but extending it to threads looked awful.  Generally speaking, if we can keep data local to the functions that creates domains or threads, with a nice block-scoped discipline (set up on entry, clean up on exit), it's always better than domain-local or thread-local storage.  

This leaves (the initial thread of) domain 0 as a special case, indeed.  I think it's best to keep the same logic we have in 4.14 (and which was unhelpfully removed when merging Multicore OCaml) concerning finalization of signal-related stuff.  There is more stuff that needs to be finalized when domain 0 exits and is not currently finalized in 5.0 (but is in 4.14), such as major heap areas.  My guess is that what we're doing with signals here paves the way to proper domain 0 finalization later in OCaml 5.

**Comment by xavierleroy** [comment]
> One aspect of this PR that is a bit fragile is the necessity to add new cleanup code on all exit points for the main domain. Other domains, and threads have clean code with a single exit point, but not so for the main domain in the cases of errors. I am not sure that all exit points have been covered.

We had this discussion already for OCaml 4 at #10726 .  I think it's better to reproduce what we did in OCaml 4 in OCaml 5 as much as possible, rather than restart from scratch.

**Review by gasche — APPROVED** [review]
I looked at this code and I believe that the change is (1) an improvement and (2) safe.

I wish we had a more structured/explicit approach to resource cleanup on thread/domain exit and also runtime exit, but this requires other changes and is not a strict requirement for the present, well-justified PR.

**Comment by lthls** [comment]
> I had a quick look, and it seems that the poll point at the start of the function triggers a GC as expected, but for some reason this GC increments the minor GC stats twice (once during `caml_empty_minor_heaps_once`, and once during `caml_major_collection_slice`; both functions are called in `caml_poll_gc_work`). I don't know yet what are the conditions for this to happen.

I took a bit more time to look at this, ans this led me to an hypothesis: the flambda version allocates slightly less (during runtime initialisation + stdlib entry code), so we manage to narrowly hit the point where the first full major GC cycle is complete during the last poll. Completing the major GC cycle triggers a minor GC, unexpectedly bumping the GC count.
And I think the reason why this PR introduces this bug might be because `caml_init_signal_handling` is not called anymore, and this means one less allocation during runtime initialisation.
If this proves correct, I think the solution is to convert all assertions in the test to use `minors_before < minors_after` instead of `minors_before+1 = minors_after`.

I'm also surprised that the only call to `caml_init_signal_handling` has disappeared, but I don't know whether it's expected or a bug.

File: runtime/domain.c
Comment (xavierleroy) [inline runtime/domain.c]: Note to reviewers: I removed this call to `caml_init_signal_handling` because there's clever code in `caml_install_signal_handler` to allocate the block of signal handlers on demand, so allocating it preventively at start-up in `caml_init_signal_handling` looked useless.  Probably I should have put that in another PR.

**Comment by xavierleroy** [comment]
Thanks @lthls for the continued investigation. I added an explanation w.r.t. `caml_init_signal_handling`.

**Comment by sadiqj** [comment]
>I think the solution is to convert all assertions in the test to use `minors_before < minors_after` instead of `minors_before+1 = minors_after`.

This is the correct solution here and preserves the property that is being tested, that there was some opportunity (via an inserted poll) to do at least one minor GC. I've created a PR for this: https://github.com/ocaml/ocaml/pull/11505

**Comment by gasche** [comment]
Thanks @lthls for the head-scratching work. I agree with the fix your proposed (and merged @sadiqj's PR), but I wanted to be sure that there wasn't a deeper issue hiding in the double counting. My understanding of your findings is that there is nothing fishy going on.

**Review by sadiqj — APPROVED** [review]
This looks good. 

I checked that signals were blocked until after we set the alternate signal stack where we do so. I also looked at our other uses threads (backup threads and the tick thread) to make sure we didn't need to do anything there. We don't as they both block all signals.

File: runtime/domain.c
Comment (xavierleroy) [reply runtime/domain.c]: I reverted this change so that we can discuss the issue separately.

**Comment by xavierleroy** [comment]
Thanks @sadiqj for the review.  I cleaned up the history and rebased on trunk.  Will merge when CI is happy.

**Comment by xavierleroy** [comment]
Cherry-picked to 5.0: eae4935100

**Comment by xavierleroy** [comment]
This PR broke the Windows port.  Emergency fix bb375e46ff seems to fix it.  Will cherry-pick to 5.0 when CI is green.

**Comment by gasche** [comment]
The problem comes from the fact that posix-signal systems have caml_init_signal_stack return NULL to indicate failure, while non-posix systems return NULL to mean a no-op (nothing to do there). Apologies for not catching this earlier in my review. A simple fix would be to have non-posix systems return a non-null pointer, for example well 1 :-)

Currently (with the hot fix) it's a bit odd that those functions are defined always but called only in some configurations. (It pushes more system-specific conditionals into the main code, which is meh.)

If the idea of returning a non-NULL pointer sounds reasonable to you, would you like me to send a PR on top of the current "hot fix" to go in this direction?

**Comment by xavierleroy** [comment]
I'm of multiple minds.  Having all signal-related stuff conditionalized with `#ifndef _WIN32` would enable more niceties, such as saving  the original signal stack at the time of `caml_init_signal_stack` and restoring it in `caml_free_signal_stack`.  So maybe that's the direction to go.

In the meantime, the "hot fix" unblocked CI on trunk, so I'll cherry-pick to 5.0 to unblock it as well.

**Pull request merged**
